forest_benchmarking/rpe.py

```diff
@@ -1,227 +1,491 @@
+    rotation = row["Rotation"]
+    meas_qubits = row["Measure Qubits"]
+    if len(meas_qubits) > 1:
+        meas_qubit = row["Non-Z-Basis Meas Qubit"]
+    else:
+        meas_qubit = meas_qubits[0]
+    post_select_state = None
+    if "Post Select State" in row.index:
+        post_select_state = row["Post Select State"]
+
+    if not isinstance(cob, Program) and qc is not None:
+        cob = change_of_basis_matrix_to_quil(qc, meas_qubits, cob)
+
+    prog = Program()
+
+    if post_select_state is None:
+        # start in equal superposition of basis states, i.e. put each qubit in plus state
+        prog = sum([local_pauli_eig_prep('X', q) for q in meas_qubits], Program())
+    else:
+        # only start the non-z-basis measurement qubit in the superposition
+        prog += local_pauli_eig_prep('X', meas_qubit)
+        # put all other qubits in the post-selection-state
+        prog += sum([RX(pi, q) for idx, q in enumerate(meas_qubits) if q != meas_qubit and
+                     post_select_state[idx] == 1], Program())
+    # using change_of_basis, transform to equal superposition of rotation eigenvectors
+    prog += cob
+    # perform the rotation depth many times
+    prog += sum([rotation for _ in range(row["Depth"])], Program())
+    # return to computational basis before measurements
+    prog += cob.dagger()
+    # prepare the meas_qubit in the appropriate meas_direction
+    prog += local_pauli_eig_meas(row["Measure Direction"], meas_qubit)
+    return prog
+
+
+def acquire_rpe_data(qc: QuantumComputer, experiments: Union[DataFrame, Sequence[DataFrame]],
+                     multiplicative_factor: float = 1.0, additive_error: float = None,
+                     grouping: Sequence[Sequence[int]] = None, results_label="Results") \
+        -> Union[DataFrame, Sequence[DataFrame]]:
+    """
+    Run each experiment in the sequence of experiments.
+
+    Each individual experiment df is copied, and raw shot outputs are stored in a column labeled by
+    results_label, which defaults to "Results". The number of shots run at each depth can be
+    modified indirectly by adjusting multiplicative_factor and additive_error.
+
+    :param experiments: dataframe containing experiments, generated by generate_rpe_experiments()
+    :param qc: a quantum computer, e.g. QVM or QPU, that runs the experiments
+    :param multiplicative_factor: ad-hoc factor to multiply the number of shots per iteration. See
+        num_trials() which computes the optimal number of shots per iteration.
+    :param additive_error: estimate of the max additive error in the experiment, see num_trials()
+    :param results_label: label for the column of the returned df to be populated with results
+    :return: A copy of the experiments data frame with the raw shot results in a new column.
+    """
+
+    if isinstance(experiments, DataFrame):
+        return run_single_rpe_experiment(qc, experiments, multiplicative_factor, additive_error,
```

Comment:
I could easily do away with run_single_rpe_experiment. I didn't because
1) it seems annoying to get a list back if you only passed in one experiment, e.g. [result_df]
2) it is easier to see what is actually going on for a single experiment. To fix 1 I could just go through all of the logic of the rest of the method, which would essentially do nothing, and in the end I check to see if len(results) is 1 and return that single value. 
I'm not sure this justifies the code duplication.

Reply:
I like it, having too much generality obscures the simplicity.